### ISSUE TITLE
About update MixedAttention operation

### ISSUE THREAD

**Issue by s9021025292140** [MCG-NJU/MixFormer#16]
Thank you for open source such an excellent work!

The original version before the update was to separate Q, K, and V into templates, online templates, and search ranges, the templates and online templates were separately performed Attention, as shown in the following:

     # template attention
        k1 = torch.cat([k_t, k_ot], dim=2)
        v1 = torch.cat([v_t, v_ot], dim=2)
        attn_score = torch.einsum('bhlk,bhtk->bhlt', [q_t, k1]) * self.scale
        attn = F.softmax(attn_score, dim=-1)
        attn = self.attn_drop(attn)
        x_t = torch.einsum('bhlt,bhtv->bhlv', [attn, v1])
        x_t = rearrange(x_t, 'b h t d -> b t (h d)')

      # online template attention
        k2 = torch.cat([k_t, k_ot], dim=2)
        v2 = torch.cat([v_t, v_ot], dim=2)
        attn_score = torch.einsum('bhlk,bhtk->bhlt', [q_ot, k2]) * self.scale
        attn = F.softmax(attn_score, dim=-1)
        attn = self.attn_drop(attn)
        x_ot = torch.einsum('bhlt,bhtv->bhlv', [attn, v2])
        x_ot = rearrange(x_ot, 'b h t d -> b t (h d)')
Especially in the calculation of attn_score, this part is calculated by q_t and k1, q_ot and k2 (both k1 and k2 are templates concatenated with online templates):

`attn_score = torch.einsum('bhlk,bhtk->bhlt', [q_t, k1]) * self.scale`

`attn_score = torch.einsum('bhlk,bhtk->bhlt', [q_ot, k2]) * self.scale`

And now the updated version is that the template and the online template are merged together to execute Attention at the beginning, as shown below:

`attn_score = torch.einsum('bhlk,bhtk->bhlt', [q_mt, k_mt]) * self.scale`

I would like to ask whether the two ways of calculating templates and online templates Attention before and after the update are equivalent?

**Comment by yutaocui** [comment]
The attention module between template and all templates (a concatenated vector of the first template and all online templates) is used to **augment** the common target in the all templates.
